library/std/src/io/buffered/bufreader.rs

```diff
@@ -236,15 +231,15 @@ impl<R: Seek> BufReader<R> {
     /// must track this information themselves if it is required.
     #[stable(feature = "bufreader_seek_relative", since = "1.53.0")]
     pub fn seek_relative(&mut self, offset: i64) -> io::Result<()> {
-        let pos = self.pos as u64;
+        let pos = self.buf.pos() as u64;
         if offset < 0 {
-            if let Some(new_pos) = pos.checked_sub((-offset) as u64) {
-                self.pos = new_pos as usize;
+            if let Some(_) = pos.checked_sub((-offset) as u64) {
+                self.buf.unconsume((-offset) as usize);
                 return Ok(());
             }
         } else if let Some(new_pos) = pos.checked_add(offset as u64) {
-            if new_pos <= self.cap as u64 {
-                self.pos = new_pos as usize;
+            if new_pos <= self.buf.cap() as u64 {
+                self.buf.consume(offset as usize);
                 return Ok(());
```

Comment:
This makes it look like we might want to return `bool` or something from consume/unconsume to indicate whether they've moved back or forward the pos marker by the passed offset.

Reply:
What do we do with the `bool` in this case? I tried replacing `consume`'s implementation with an `assert!` that the increment doesn't exceed the filled region of the buffer and just doing the addition, and didn't see the optimization come back. I just can't think of something else to do with the `bool` other than assert that it is `false` when we expect it to be.

Reply:
I was thinking these checks would be inlined, so we'd have something like this, where the checked_sub/checked_add happens inside unconsume/consume. 

```rust
if offset < 0 {
    if self.buf.unconsume((-offset) as usize) {
        return Ok(());
    }
} else {
    if self.buf.consume(offset as usize) {
        return Ok(());
    }
}
```

Reply:
Oh! This code is not related to the optimization I was looking for. The interesting code is all in `read_exact`. And as far as I can tell, the only way to recover the optimization is to do unconditionally increment `pos` by the amount of bytes read in that function.

The change you're suggesting adds complexity to the code path that I would like to see optimized because then `consume` would have to check for overflow _and_ going out of the initialized region. I don't have any benchmarks where seeking in a `BufReader` is hot, but if you do I'm happy to take a look.

Reply:
Ah, OK. In that case this seems fine :)

I think instead of consume_unchecked we might be able to do that by having consume return the consumed buffer (`Option<&[u8]>`)? I think that should hopefully optimize away when we don't need it (i.e., in seeking), but potentially allows LLVM to only do the comparisons necessary once, and avoids any consume_unchecked?

Reply:
I couldn't figure out how to implement a `try_consume -> Option<&[u8]>` without adding more unsafe, because calling `self.buffer()` borrows `self` so `self.pos` can't be updated. So we'd need the internals of `Buffer::buffer` in two places.

But a closure-style method works great here and gets the optimization I'm looking for.